### PR TITLE
0-dim batch size input for interpolate.

### PR DESCRIPTION
**Pull request by v0dro**
This PR adds support for 0-dim batch size input for `torch.nn.functional.interpolate` for various modes of interpolation.

Fixes part of gh-12013

CC: @rgommers  @ezyang

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit def74d00:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=58234c0254f135d1e3714b17f44f4922d56d41b4&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base 58234c02 since Jan 18
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=def74d000c187b8b3799c2b1dc6348720f6f9d69).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_linux_backward_compatibility_check_test](https://dr.pytorch.org/build-details.html?build_id=88350801) from Jan 18



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: aten/src/ATen/native/UpSample.h [hunk 1]
```diff
@@ -60,10 +60,15 @@ static inline void upsample_1d_shape_check(
...
+    valid_empty = input.size(0) == 0 && input.size(1) != 0 && input.size(2) != 0;
+    
     TORCH_CHECK(
-        input.numel() != 0 && input.dim() == 3,
-        "Non-empty 3D data tensor expected but got a tensor with sizes ",
-        input.sizes());
+                (input.numel() != 0 || valid_empty) && input.dim() == 3,
+                "Non-empty 3D data tensor expected but got a tensor with sizes ",
+                input.sizes());
+    std::cout << "post input check.\n";
```
Comment (ezyang) [inline aten/src/ATen/native/UpSample.h]: You don't want this

File: aten/src/ATen/native/UpSample.h [hunk 2]
```diff
@@ -60,10 +60,15 @@ static inline void upsample_1d_shape_check(
...
   if (input.defined()) {
+    // Allow for empty batch size but not other dimensions
+    bool valid_empty = false;
+    valid_empty = input.size(0) == 0 && input.size(1) != 0 && input.size(2) != 0;
+    
     TORCH_CHECK(
-        input.numel() != 0 && input.dim() == 3,
-        "Non-empty 3D data tensor expected but got a tensor with sizes ",
-        input.sizes());
+                (input.numel() != 0 || valid_empty) && input.dim() == 3,
```
Comment (ezyang) [inline aten/src/ATen/native/UpSample.h]: This seems like an awfully circuitous way to write the condition. Since dimensionality is known, just assert that `input.size(1) != 0 && input.size(2) != 0` and call it a day.

**Review by ezyang — CHANGES_REQUESTED** [review]
Looks reasonable but some nits.

File: aten/src/ATen/native/UpSample.h [hunk 3]
```diff
@@ -60,10 +60,12 @@ static inline void upsample_1d_shape_check(
       ")");
 
   if (input.defined()) {
+    // Allow for empty batch size but not other dimensions
     TORCH_CHECK(
-        input.numel() != 0 && input.dim() == 3,
-        "Non-empty 3D data tensor expected but got a tensor with sizes ",
-        input.sizes());
+                (input.numel() != 0 ||
```
Comment (ezyang) [inline aten/src/ATen/native/UpSample.h]: The test on numel is totally unnecessary now... because if a[1] and a[2] are nonzero, numel can only be zero when a[0] is zero ;)

File: aten/src/ATen/native/UpSample.h [hunk 4]
```diff
@@ -136,10 +142,14 @@ static inline void upsample_3d_shape_check(
       ")");
 
   if (input.defined()) {
+    // Allow for empty batch size but not other dimensions
+    bool valid_empty = false;
+    valid_empty = input.size(0) == 0 && input.size(1) != 0 &&
+      input.size(2) != 0 && input.size(3) != 0 && input.size(4) != 0;
```
Comment (ezyang) [inline aten/src/ATen/native/UpSample.h]: nit: just assign the expression directly to `valid_empty`, there's no reason to assign it to false first and then reassign

**Review by ezyang — APPROVED** [review]
